Adding scenario for multinode job array with ASAP and standing reservation

Mahalty opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Description
Automate following scenario 

- Multinode job array with ASAP reservation

- Multinode job array with standing reservation


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
[Execution_logs_resv.txt](https://github.com/PBSPro/pbspro/files/3334456/Execution_logs_resv.txt)



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+
+    def test_ASAP_resv_with_multinode_job_array(self):
+        """
+        Test multinode job array with ASAP reservation
```
Can you please explain why do you need multi-node jobs? This test isn't testing anything related to multiple nodes. Consider making it multi-vnode test instead.
Also, in this test you are submitting two different array jobs, creating 2 ASAP reservations out of it. Can you please write a docstring about what you are trying to test with multiple array jobs and reservations that can not be tested with one.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+                 'Resource_List.walltime': '30',
+                 'Resource_List.place': 'scatter'}
+        j = Job(PBSROOT_USER)
+        j.set_sleep_time(30)
```
Please consider reducing the sleeptime and walltime to something smaller (like 5 or 10 seconds) 30 seconds is too long

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        attrs = {ATTR_J: '1-5',
+                 'Resource_List.select': '1:ncpus=1',
+                 'Resource_List.walltime': '30',
+                 'Resource_List.place': 'scatter'}
```
it is a single chunk job, scatter placement does not mean anything here

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        Test multinode job array with standing reservation
+        """
+        if (len(self.moms) < 2):
+            self.skip_test("Test requires 2 moms: use -p mom1:mom2")
```
Same comment as above, why do we need 2 moms? Can you use 2 vnodes?

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        self.mom.isUp()
+        self.scheduler.isUp()
+
+    @timeout('1600')
```
Why do you have a 1600 seconds timeout on this test?

Mahalty commented:
Latest execution logs;
[Latest_execution_logs_TestReservations.txt](https://github.com/PBSPro/pbspro/files/3363263/Latest_execution_logs_TestReservations.txt)

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1150,16 +1158,15 @@ def test_ASAP_resv_with_multinode_job_array(self):
-                 'Resource_List.place': 'scatter'}
+                 'Resource_List.select': '2:ncpus=1',
+                 'Resource_List.walltime': '15',
+                 'Resource_List.place': 'vscatter'}
```
We can remove line 1164 to 1167.

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1124,23 +1124,31 @@ def test_reservations_with_expired_subjobs(self):
+        self.server.create_vnodes("vnode1", vn_attrs, 1,
+                                  self.mom, fname="vnodedef1")
+        self.server.create_vnodes("vnode2", vn_attrs, 1,
+                                  self.mom, additive=True, fname="vnodedef2")
```
Can you please check if we can create vnode as vn1[0] and vn1[1] instead of vn1[0] and vn2[0]?
If yes, then  delete line 1136 and 1137.

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+                           extend='t', id=jid)
+        self.server.expect(RESV, {'reserve_index': 2}, offset=60, id=rid)
+        self.server.expect(JOB, {'job_state=R': 2}, count=True,
+                           extend='t', id=jid)
```
Here we can add a check to verify , J1[5] should be in Q state.

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        self.server.expect(RESV, {'reserve_index': 2}, offset=60, id=rid)
+        self.server.expect(JOB, {'job_state=R': 2}, count=True,
+                           extend='t', id=jid)
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid, offset=60)
```
This operation we could perform once reservation deleted from server.

Mahalty commented:
@Klovely  Thanks for the review. I have addressed your review comments. Please review
Latest execution logs:
[Execution_logs_TestReservations_1.txt](https://github.com/PBSPro/pbspro/files/3368171/Execution_logs_TestReservations_1.txt)

Mahalty commented:
@arungrover @Klovely : I have addressed your review comments and attached latest execution logs. Please review the latest commit

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1244,6 +1242,9 @@ def test_standing_resv_with_multinode_job_array(self):
                            extend='t', id=jid)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=subjid[4])
         self.server.expect(JOB, 'queue', op=UNSET, id=jid, offset=60)
+        self.server.log_match(resv_queue + ";deleted at request of pbs_server",
```
Only one comment 
This line should come before self.server.expect(JOB, 'queue', op=UNSET, id=jid, offset=60)
Rest looks fine.

Mahalty commented:
Thanks @Klovely for the review. I have addressed your review comment and below is the latest execution logs.
Latest execution-logs:
[Latest_TestReservations.txt](https://github.com/PBSPro/pbspro/files/3376879/Latest_TestReservations.txt)

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+
+    def test_ASAP_resv_with_multivnode_job_array(self):
+        """
+        Test multivnode job array with ASAP reservation. Also
```
I personally think this test is not needed. Lovely's test would cover the scenario where array jobs are submitted to reservations (advance or standing), I'm not sure what we gain from submitting multi-vnode (or multi-node) jobs in this test suite.

Mahalty replied:
@arungrover This test was added as part of regression testing.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        j1 = Job(PBSROOT_USER)
+        j1.set_sleep_time(15)
+        j1.set_attributes(attrs)
+        jid2 = self.server.submit(j1)
```
Please write a comment about why you are submitting second array job and how this job will help you get the same start time as the previous ASAP reservation

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+                           extend='xt', id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, subjid[4])
+        self.server.expect(JOB, {'job_state': 'R'}, subjid2[0])
+        # Wait for reservation to delete from server
```
Please change it to "wait for reservation to end".

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+
+        start = int(time.time()) + 10
+        end = int(time.time()) + 61
+        a = {'Resource_List.select': '2:ncpus=2',
```
Since there are 2 vnodes with 4 cpus each, it may happen that both the chunks of this reservation get satisfied on the same vnode, in that case, it wouldn't be multi-vnode as you wanted it to be. Use place=vscatter

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+        resv_queue = rid.split(".")[0]
+
+        attrs = {ATTR_q: resv_queue, ATTR_J: '1-5',
+                 'Resource_List.select': '2:ncpus=1'}
```
Add place=vscatter

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,128 @@ def test_reservations_with_expired_subjobs(self):
+
+    def test_standing_resv_with_multivnode_job_array(self):
+        """
+        Test multivnode job array with standing reservation
```
You are not look for server/sched/mom to be up in this test. Please write the intent behind writing this test.

Mahalty commented:
**Latest execution-logs:**
[Latest_logs_TestReservations.txt](https://github.com/PBSPro/pbspro/files/3396793/Latest_logs_TestReservations.txt)

Mahalty commented:
Latest execution results:
[Execution_logs_TestReservations_latest.txt](https://github.com/PBSPro/pbspro/files/3416367/Execution_logs_TestReservations_latest.txt)
